src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1395,16 +1395,23 @@ public SolutionCompilationState WithFrozenSourceGeneratedDocuments(
         using var _ = PooledDictionary<DocumentId, SourceGeneratedDocumentState>.GetInstance(out var documentStates);
         foreach (var (documentIdentity, generationDateTime, sourceText, syntaxNode) in documents)
         {
+            Contract.ThrowIfTrue(sourceText is null && syntaxNode is null);
+            Contract.ThrowIfTrue(sourceText is not null && syntaxNode is not null);
```

Comment:
I like this generally -- it's clear we're always doing exactly one or the other.